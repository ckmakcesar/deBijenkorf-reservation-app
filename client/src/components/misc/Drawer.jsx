import React, { useRef, useEffect } from 'react';
import clsx from 'clsx';
import PropTypes from 'prop-types';

import Header from './Header';
import Footer from './Footer';
import Button from './Button';
import { isEscKeyDown } from '../../utils/utils';

import styles from '../../styles/misc/Drawer.module.css';

const Drawer = ({
  isOpen,
  headerTitle,
  formId,
  confirmButtonText,
  confirmButtonTone,
  onClose,
  onConfirm,
  id,
  children,
}) => {
  const bodyRef = useRef(document.querySelector('body'));

  useEffect(() => {
    const updatePageScroll = () => {
      if (isOpen) {
        bodyRef.current.style.overflow = 'hidden';
      } else {
        bodyRef.current.style.overflow = '';
      }
    };
    updatePageScroll();
  }, [isOpen]);

  const handleSubmit = (payload) => {
    onConfirm(payload);
    onClose();
  };

  const confirmButtonOptionalProps = {};
  if (formId) {
    confirmButtonOptionalProps.isSubmit = true;
    confirmButtonOptionalProps.formId = formId;
  } else {
    confirmButtonOptionalProps.onClick = onConfirm;
  }

  // set focus to the overlay, and quit if ESCAPE is pressed
  const focusDiv = useRef();
  useEffect(() => {
    if (isOpen && focusDiv.current) {
      focusDiv.current.focus();
    }
  }, [isOpen]);
  const handleKeyDown = (e) => {
    if (isEscKeyDown(e)) {
      onClose();
    }
  };

  return (
    <div
      {...(id ? { id: `Drawer-${id}` } : null)} // optional id - should be unique
      className={styles.drawerContainer}
      ref={focusDiv}
      tabIndex={0}
      onKeyDown={handleKeyDown}
    >
      <div className={clsx(styles.drawer, isOpen && styles.openedDrawer)}>
        <Header
          {...(id ? { id: `${id}-drawer` } : null)} // optional id - should be unique
          text={headerTitle}
        />

        {/* drawer content -> flex & overflow scrol */}
        <div className='flex-content'>
          {React.cloneElement(
            children,
            {
              formId,
              onSubmit: handleSubmit,
              onClose,
            }
          )}
        </div>

        <Footer
          {...(id ? { id: `${id}-drawer` } : null)} // optional id - should be unique
        >
          <div className={styles.drawerButtons}>
            <Button
              {...(id ? { id: `${id}-dialog-cancel` } : null)} // optional id - should be unique
              className='flex-grow'
              onClick={() => { onClose() }}
              text='Cancel'
              tone='secondary'
            />

            <Button
              {...(id ? { id: `${id}-drawer-confirm` } : null)} // optional id - should be unique
              className='flex-grow'
              text={confirmButtonText}
              tone={confirmButtonTone}
              {...confirmButtonOptionalProps}
            />
          </div>
        </Footer>
      </div>
    </div>
  );
}

// *********************************************************************************************************************
// PROPS
// *********************************************************************************************************************
Drawer.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  headerTitle: PropTypes.string.isRequired,
  formId: PropTypes.string,
  confirmButtonText: PropTypes.string,
  confirmButtonTone: PropTypes.string,
  onClose: PropTypes.func.isRequired,
  onConfirm: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
  id: PropTypes.string,
};

Drawer.defaultProps = {
  formId: '',
  confirmButtonText: 'Confirm',
  confirmButtonTone: '',
  id: '',
};

export default Drawer;