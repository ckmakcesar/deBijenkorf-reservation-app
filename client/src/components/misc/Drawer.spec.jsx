import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import Drawer from './Drawer';

describe('component:Drawer', () => {
  it('Should render Drawer with every optional props absent', () => {
    const wrapper = shallow(
      <Drawer
        isOpen={true}
        headerTitle='Top of Page'
        onClose={jest.fn()}
        onConfirm={jest.fn()}
      >
        {/* lint recommends NOT to nest children as prop */}
        <h5>hello world</h5>
      </Drawer>
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('Should render Drawer with every optional props present', () => {
    const wrapper = shallow(
      <Drawer
        isOpen={true}
        headerTitle='Top of Page'
        formId='form-002'
        confirmButtonText='Vamos Amigo'
        confirmButtonTone='green'
        onClose={jest.fn()}
        onConfirm={jest.fn()}
      >
        <h5>hello world</h5>
      </Drawer>
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
