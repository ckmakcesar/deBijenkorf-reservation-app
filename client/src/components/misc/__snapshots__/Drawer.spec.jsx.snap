// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component:Drawer Should render Drawer with every optional props absent 1`] = `
<div
  className="drawerContainer"
  onKeyDown={[Function]}
  tabIndex={0}
>
  <div
    className="drawer openedDrawer"
  >
    <Header
      id=""
      smaller={false}
      text="Top of Page"
    />
    <div
      className="flex-content"
    >
      <h5
        formId=""
        onClose={[MockFunction]}
        onSubmit={[Function]}
      >
        hello world
      </h5>
    </div>
    <Footer
      id=""
      reducedPadding={false}
    >
      <div
        className="drawerButtons"
      >
        <Button
          boxIconClassName=""
          className="flex-grow"
          disabled={false}
          formId=""
          id=""
          isSubmit={false}
          onClick={[Function]}
          text="Cancel"
          tone="secondary"
        />
        <Button
          boxIconClassName=""
          className="flex-grow"
          disabled={false}
          formId=""
          id=""
          isSubmit={false}
          onClick={[MockFunction]}
          text="Confirm"
          tone=""
        />
      </div>
    </Footer>
  </div>
</div>
`;

exports[`component:Drawer Should render Drawer with every optional props present 1`] = `
<div
  className="drawerContainer"
  onKeyDown={[Function]}
  tabIndex={0}
>
  <div
    className="drawer openedDrawer"
  >
    <Header
      id=""
      smaller={false}
      text="Top of Page"
    />
    <div
      className="flex-content"
    >
      <h5
        formId="form-002"
        onClose={[MockFunction]}
        onSubmit={[Function]}
      >
        hello world
      </h5>
    </div>
    <Footer
      id=""
      reducedPadding={false}
    >
      <div
        className="drawerButtons"
      >
        <Button
          boxIconClassName=""
          className="flex-grow"
          disabled={false}
          formId=""
          id=""
          isSubmit={false}
          onClick={[Function]}
          text="Cancel"
          tone="secondary"
        />
        <Button
          boxIconClassName=""
          className="flex-grow"
          disabled={false}
          formId="form-002"
          id=""
          isSubmit={true}
          onClick={[Function]}
          text="Vamos Amigo"
          tone="green"
        />
      </div>
    </Footer>
  </div>
</div>
`;
